contribs/src/main/scala/io/delta/storage/BaseExternalLogStore.scala

```diff
@@ -0,0 +1,407 @@
+      )
+      .values
+      .toSeq
+      .sortBy(_.getPath.getName)
+    result.iterator
+  }
+
+  private def resolved(path: Path): (FileSystem, Path) = {
+    val fs = path.getFileSystem(getHadoopConfiguration)
+    val resolvedPath = stripUserInfo(fs.makeQualified(path))
+    (fs, resolvedPath)
+  }
+
+  private def stripUserInfo(path: Path): Path = {
+    val uri = path.toUri
+    val newUri = new URI(
+      uri.getScheme,
+      null,
+      uri.getHost,
+      uri.getPort,
+      uri.getPath,
+      uri.getQuery,
+      uri.getFragment
+    )
+    new Path(newUri)
+  }
+
+  /**
+   * Best-effort at ensuring consistency on file system according to the external cache.
+   *
+   * Tries to copy a LogEntryMetadata's temp file into the target _delta_log file, if it does not
+   * exist already, and then commit to the external cache.
+   *
+   * Will retry at most 3 times.
+   *
+   * @return the correct FileStatus from which to read the entry's data
+   */
+  private def tryFixTransactionLog(
+      fs: FileSystem,
+      entry: LogEntryMetadata,
+      failureAcceptable: Boolean): FileStatus = {
+    logDebug(s"Try to fix: ${entry.path}")
+    val completedEntry = entry.complete()
+    var fileCopied = false
+
+    for (i <- 0 until 3) {
+      try {
+        if (!fileCopied && !fs.exists(entry.path)) {
+          copyFile(fs, entry.tempPath, entry.path)
+          fileCopied = true
+        }
+
+        // Since overwrite is true, we do not expect any FileAlreadyExistsExceptions
+        writeCache(fs, completedEntry, overwrite = true)
+
+        return completedEntry.asFileStatus(fs)
+      } catch {
+        case NonFatal(e) =>
+          logWarning(s"${e.getClass.getName}: Ignoring error while fixing. Iteration $i of 3. $e")
+      }
+    }
+
+    if (failureAcceptable) {
+      // We weren't able to commit to the external cache, but we can still read from the temp file
+      entry.tempPathAsFileStatus(fs)
+    } else {
+      throw new RuntimeException(s"Unable to fix the transaction log for path ${entry.path}")
+    }
+  }
+
+  /**
+   * Returns path stripped user info.
+   */
+  protected def getPathKey(resolvedPath: Path): Path = {
+    stripUserInfo(resolvedPath)
+  }
+
+  /**
+   * Generate temporary path for TransactionLog.
+   */
+  protected def getTemporaryPath(path: Path): Path = {
+    val uuid = java.util.UUID.randomUUID().toString
+    new Path(s"${path.getParent}/.temp/${path.getName}.$uuid")
+  }
+
+  /**
+   * List files starting from `resolvedPath` (inclusive) in the same directory, which merges
+   * the file system list and the db list
+   */
+  protected def listFrom(
+      fs: FileSystem,
+      resolvedPath: Path,
+      useCache: Boolean = true): Iterator[FileStatus] = {
+    val parentPath = resolvedPath.getParent
+    if (!fs.exists(parentPath)) {
+      throw new FileNotFoundException(s"No such file or directory: $parentPath")
+    }
+
+    val listedFromFs =
```

Comment:
One idea:
- this is a `lazy val`
- `listFromExternal` returns a sorted iterator

We can then listFromExternal, and THEN use the `listedFromFS` variable (which was lazy so it was just executed, and it would see the latest FS files). Then there's no need to merge file iterators. the FS will have all the latest

Reply:
But if the copying fails ... then we will have to use the temp file ?

So perhaps that's what we do. We only return files that FAILED to get copied into N.json in the `listedFromDB` iterator (then this is a much smaller list of files.). Then almost all of the files will be in `listedFromFS` iterator.